src/ipld/util.rs

```diff
@@ -595,19 +594,22 @@ impl<DB: Blockstore + Send + Sync + 'static, T: Iterator<Item = Tipset> + Unpin>
                         }
                     }
                 }
-            } else if let Ok(item) = this.block_receiver.try_recv() {
-                if let Some(item) = item {
-                    return Poll::Ready(Some(item));
-                }
-                // Close the sender when it's empty and exit.
-                if this.extract_sender.is_empty() && this.block_receiver.is_empty() {
-                    this.worker_handle.abort();
-                    return Poll::Ready(None);
-                }
             } else {
-                // That's it, nothing else to do. End of stream.
-                return Poll::Ready(None);
-            };
+                match this.block_receiver.try_recv() {
+                    Ok(item) => return Poll::Ready(Some(item)),
+                    Err(err) => {
+                        if this.extract_sender.is_empty() {
+                            this.worker_handle.abort();
+                            return Poll::Ready(None);
+                            // This should never happen.
+                        } else if err == TryRecvError::Disconnected {
+                            panic!(
```

Comment:
This should never happen or can' happen at all? Panicking is rough and we wouldn't like Forest crashing under certain conditions.

Reply:
Otherwise we could log an error and end the stream, which I reckon isn’t the most amazing solution to this problem.

The stream interface does not cater for errors, so there isn’t much of a choice.

Reply:
Let's have a comment on why this cannot happen, given it's not obvious (at least to me). If this ever happens, we should at least know what expectations didn't hold.

Reply:
done

Reply:
@ruseinov where?

Reply:
oops, it didn't push because of the branch updates. I'll push a new PR.

Reply:
https://github.com/ChainSafe/forest/pull/3928